File: core/src/main/java/dev/vibris/core/CapturePlanBuilder.java
package dev.vibris.core;

import dev.vibris.api.CapturePlan;
import dev.vibris.api.ResourceCatalog;
import dev.vibris.protocol.v1.Action;
import dev.vibris.protocol.v1.ErrorCode;

import java.util.ArrayList;
import java.util.HashSet;
import java.util.List;
import java.util.Locale;
import java.util.Set;

final class CapturePlanBuilder {
    private static final int MAX_ACTIONS = 64;

    int waitFrames(CoreJob job) throws RuntimeJobExecutor.Failure {
        try {
            requireActionLimit(job);
            long frames = job.submission.hasRecipe() ? recipeFrames(job) : actionFrames(job);
            if (frames > Integer.MAX_VALUE) {
                throw new ArithmeticException("frame count");
            }
            return (int) frames;
        } catch (ArithmeticException exception) {
            throw new RuntimeJobExecutor.Failure(ErrorCode.INTERNAL_ERROR, "Requested frame count is too large.");
        }
    }

    Plan build(CoreJob job, ResourceCatalog catalog) throws RuntimeJobExecutor.Failure {
        try {
            List<CapturePlan.Target> targets = new ArrayList<>();
            if (job.submission.hasActions()) {
                for (Action action : job.submission.getActions().getActionsList()) addAction(targets, action, catalog);
            } else if (job.submission.getRecipe().hasReloadAndCapture()) {
                var recipe = job.submission.getRecipe().getReloadAndCapture();
                targets.add(screenshot(catalog, format(recipe.getScreenshotFormat(), CapturePlan.ArtifactFormat.PNG),
                    "screenshot"));
            } else if (job.submission.getRecipe().hasCaptureDebugBundle()) {
                addDebugBundle(targets, job, catalog);
            }
            if (targets.size() > MAX_ACTIONS) {
                throw new RuntimeJobExecutor.Failure(ErrorCode.CAPTURE_FAILED, "Capture target limit exceeded.");
            }
            CapturePlan plan = new CapturePlan(targets);
            return new Plan(plan, validateAndEstimate(plan, catalog));
        } catch (IllegalArgumentException exception) {
            throw new RuntimeJobExecutor.Failure(ErrorCode.CAPTURE_FAILED, "Capture plan is invalid.");
        }
    }

    private static void addDebugBundle(List<CapturePlan.Target> targets, CoreJob job, ResourceCatalog catalog) {
        var recipe = job.submission.getRecipe().getCaptureDebugBundle();
        if (recipe.getScreenshot()) targets.add(screenshot(catalog, CapturePlan.ArtifactFormat.PNG, "screenshot"));
        recipe.getTexturesList().forEach(name -> targets.add(target(
            ResourceCatalog.ResourceKind.TEXTURE, name, CapturePlan.ArtifactFormat.RAW, name, 0, 0)));
        recipe.getBuffersList().forEach(name -> targets.add(target(
            ResourceCatalog.ResourceKind.BUFFER, name, CapturePlan.ArtifactFormat.BIN, name, 0, 0)));
    }

    private static void addAction(List<CapturePlan.Target> targets, Action action, ResourceCatalog catalog) {
        if (action.hasCaptureScreenshot()) {
            var capture = action.getCaptureScreenshot();
            targets.add(screenshot(catalog, format(capture.getFormat(), CapturePlan.ArtifactFormat.PNG),
                capture.getArtifactName().isBlank() ? "screenshot" : capture.getArtifactName()));
        } else if (action.hasDumpTexture()) {
            var dump = action.getDumpTexture();
            targets.add(target(ResourceCatalog.ResourceKind.TEXTURE, dump.getLogicalName(),
                format(dump.getFormat(), CapturePlan.ArtifactFormat.RAW), dump.getArtifactName(),
                dump.getMipLevel(), dump.getLayer()));
        } else if (action.hasDumpBuffer()) {
            var dump = action.getDumpBuffer();
            targets.add(target(ResourceCatalog.ResourceKind.BUFFER, dump.getLogicalName(),
                format(dump.getFormat(), CapturePlan.ArtifactFormat.BIN), dump.getArtifactName(), 0, 0));
        }
    }

    private static CapturePlan.Target screenshot(ResourceCatalog catalog, CapturePlan.ArtifactFormat format,
        String artifactName) {
        String name = catalog.resources().stream()
            .filter(resource -> resource.kind() == ResourceCatalog.ResourceKind.FINAL_FRAMEBUFFER)
            .map(ResourceCatalog.ResourceDescriptor::logicalName).findFirst().orElse("final_framebuffer");
        return target(ResourceCatalog.ResourceKind.FINAL_FRAMEBUFFER, name, format, artifactName, 0, 0);
    }

    private static CapturePlan.Target target(ResourceCatalog.ResourceKind kind, String name,
        CapturePlan.ArtifactFormat format, String artifactName, int mip, int layer) {
        return new CapturePlan.Target(kind, name, format, artifactName, mip, layer);
    }

    private static long validateAndEstimate(CapturePlan plan, ResourceCatalog catalog)
        throws RuntimeJobExecutor.Failure {
        long bytes = 0;
        Set<String> names = new HashSet<>();
        try {
            for (CapturePlan.Target target : plan.targets()) {
                if (!supported(target)) {
                    throw new RuntimeJobExecutor.Failure(
                        ErrorCode.CAPTURE_FAILED, "Capture resource kind and format are incompatible.");
                }
                if (!names.add(canonical(target.fileName())) || !names.add(canonical(target.metadataFileName()))) {
                    throw new RuntimeJobExecutor.Failure(
                        ErrorCode.CAPTURE_FAILED, "Capture artifact names are repeated.");
                }
                ResourceCatalog.ResourceDescriptor resource = find(catalog, target);
                if (target.mipLevel() >= Math.max(1, resource.mipLevels()) ||
                    target.layer() >= Math.max(1, resource.layers())) {
                    throw missing(target.logicalName());
                }
                bytes = Math.addExact(bytes, resource.byteSize());
            }
            return bytes;
        } catch (ArithmeticException exception) {
            throw new RuntimeJobExecutor.Failure(
                ErrorCode.ARTIFACT_JOB_TOO_LARGE, "Artifact estimate is too large.");
        }
    }

    private static ResourceCatalog.ResourceDescriptor find(ResourceCatalog catalog, CapturePlan.Target target)
        throws RuntimeJobExecutor.Failure {
        return catalog.resources().stream()
            .filter(resource -> resource.kind() == target.kind() && resource.logicalName().equals(target.logicalName()))
            .findFirst().orElseThrow(() -> missing(target.logicalName()));
    }

    private static RuntimeJobExecutor.Failure missing(String name) {
        return new RuntimeJobExecutor.Failure(
            ErrorCode.CAPTURE_RESOURCE_NOT_FOUND, "Capture resource was not found: " + name);
    }

    private static long actionFrames(CoreJob job) {
        long frames = 0;
        for (Action action : job.submission.getActions().getActionsList()) {
            if (action.hasWaitFrames()) frames = Math.addExact(frames, action.getWaitFrames().getFrameCount());
        }
        return frames;
    }

    private static void requireActionLimit(CoreJob job) throws RuntimeJobExecutor.Failure {
        if (job.submission.hasActions() && job.submission.getActions().getActionsCount() > MAX_ACTIONS) {
            throw new RuntimeJobExecutor.Failure(ErrorCode.CAPTURE_FAILED, "Action limit exceeded.");
        }
    }

    private static String canonical(String name) {
        return name.toLowerCase(Locale.ROOT);
    }

    private static boolean supported(CapturePlan.Target target) {
        return switch (target.kind()) {
            case FINAL_FRAMEBUFFER -> target.format() == CapturePlan.ArtifactFormat.PNG;
            case TEXTURE -> target.format() == CapturePlan.ArtifactFormat.PNG ||
                target.format() == CapturePlan.ArtifactFormat.RAW;
            case BUFFER -> target.format() == CapturePlan.ArtifactFormat.BIN;
        };
    }

    private static long recipeFrames(CoreJob job) {
        var recipe = job.submission.getRecipe();
        if (recipe.hasReloadAndCapture()) return recipe.getReloadAndCapture().getWarmupFrames();
        if (recipe.hasCaptureDebugBundle()) return recipe.getCaptureDebugBundle().getWarmupFrames();
        if (recipe.hasAbCompare()) return recipe.getAbCompare().getWarmupFrames();
        return 0;
    }

    private static CapturePlan.ArtifactFormat format(dev.vibris.protocol.v1.ArtifactFormat format,
        CapturePlan.ArtifactFormat fallback) {
        return switch (format) {
            case ARTIFACT_FORMAT_PNG -> CapturePlan.ArtifactFormat.PNG;
            case ARTIFACT_FORMAT_EXR -> CapturePlan.ArtifactFormat.EXR;
            case ARTIFACT_FORMAT_RAW -> CapturePlan.ArtifactFormat.RAW;
            case ARTIFACT_FORMAT_BIN -> CapturePlan.ArtifactFormat.BIN;
            default -> fallback;
        };
    }

    record Plan(CapturePlan capture, long estimatedBytes) {
    }
}

File: core/src/main/kotlin/dev/vibris/core/CapturePlanBuilder.kt
package dev.vibris.core

import dev.vibris.api.CapturePlan
import dev.vibris.api.ResourceCatalog
import dev.vibris.protocol.v2.Action
import dev.vibris.protocol.v2.ErrorCode
import java.util.Locale

internal object CapturePlanBuilder {
    @JvmRecord
    data class Plan(val capture: CapturePlan, val estimatedBytes: Long)

    fun addAction(
        targets: MutableList<CapturePlan.Target>,
        action: Action,
        catalog: ResourceCatalog,
    ) {
        when {
            action.hasTakeScreenshot() -> {
                val capture = action.takeScreenshot
                targets.add(
                    screenshot(
                        catalog,
                        format(capture.format, CapturePlan.ArtifactFormat.PNG),
                        capture.artifactName.ifBlank { "screenshot" },
                    ),
                )
            }
            action.hasDumpTexture() -> {
                val dump = action.dumpTexture
                val selector = dump.resource
                targets.add(
                    target(
                        ResourceCatalog.ResourceKind.TEXTURE,
                        selector.logicalName,
                        textureFormat(dump.format),
                        dump.artifactName,
                        selector.mipLevel,
                        selector.layer,
                    ),
                )
            }
            action.hasDumpBuffer() -> {
                val dump = action.dumpBuffer
                targets.add(
                    target(
                        ResourceCatalog.ResourceKind.BUFFER,
                        dump.logicalName,
                        CapturePlan.ArtifactFormat.BIN,
                        dump.artifactName,
                        0,
                        0,
                    ),
                )
            }
        }
    }

    @Throws(RuntimeJobExecutor.Failure::class)
    fun plan(targets: List<CapturePlan.Target>, catalog: ResourceCatalog): Plan {
        val capture = CapturePlan(targets.map { expand(it, catalog) })
        return Plan(capture, validateAndEstimate(capture, catalog))
    }

    fun screenshot(
        catalog: ResourceCatalog,
        format: CapturePlan.ArtifactFormat,
        artifactName: String,
    ): CapturePlan.Target {
        val name = catalog.resources
            .firstOrNull { it.kind == ResourceCatalog.ResourceKind.FINAL_FRAMEBUFFER }
            ?.logicalName
            ?: "final_framebuffer"
        return target(
            ResourceCatalog.ResourceKind.FINAL_FRAMEBUFFER,
            name,
            format,
            artifactName,
            0,
            0,
        )
    }

    fun target(
        kind: ResourceCatalog.ResourceKind,
        name: String,
        format: CapturePlan.ArtifactFormat,
        artifactName: String,
        mip: Int,
        layer: Int,
    ): CapturePlan.Target = CapturePlan.Target(kind, name, format, artifactName, mip, layer)

    fun patchedShaders(artifactName: String): CapturePlan = CapturePlan(listOf(CapturePlan.Target(
        ResourceCatalog.ResourceKind.PATCHED_SHADERS,
        "patched_shaders",
        CapturePlan.ArtifactFormat.TEXT,
        artifactName,
        0,
        0,
    )))

    fun realizePatchedShaders(plan: CapturePlan, result: dev.vibris.api.CaptureResult): CapturePlan {
        val target = plan.targets.single()
        require(target.kind == ResourceCatalog.ResourceKind.PATCHED_SHADERS && target.outputs.isEmpty()) {
            "Patched shader capture plan is invalid"
        }
        val group = result.groups.single()
        require(group.name == target.artifactName &&
            group.resource.kind == target.kind && group.resource.logicalName == target.logicalName) {
            "Patched shader capture result did not match its plan"
        }
        val outputs = group.artifacts.map { artifact ->
            CapturePlan.ArtifactOutputSpec(
                artifact.fileName,
                artifact.format,
                artifact.role,
                artifact.subresourceIndex,
            )
        }
        require(outputs.map { canonical(it.fileName) }.toSet().size == outputs.size) {
            "Patched shader artifact names are repeated"
        }
        return CapturePlan(listOf(target.copy(outputs = outputs)))
    }

    @Throws(RuntimeJobExecutor.Failure::class)
    private fun validateAndEstimate(plan: CapturePlan, catalog: ResourceCatalog): Long {
        var bytes = 0L
        val names = HashSet<String>()
        try {
            for (target in plan.targets) {
                if (!supported(target)) {
                    throw RuntimeJobExecutor.Failure(
                        ErrorCode.ERROR_CODE_CAPTURE_FAILED,
                        "Capture resource kind and format are incompatible.",
                    )
                }
                for (output in target.outputs) {
                    if (!names.add(canonical(output.fileName))) {
                        throw RuntimeJobExecutor.Failure(
                            ErrorCode.ERROR_CODE_CAPTURE_FAILED,
                            "Capture artifact names are repeated.",
                        )
                    }
                }
                val resource = catalog.resources
                    .firstOrNull { it.kind == target.kind && it.logicalName == target.logicalName }
                    ?: throw missing(target.logicalName)
                if (
                    target.mipLevel >= maxOf(1, resource.mipLevels) ||
                    target.layer >= maxOf(1, resource.layers)
                ) {
                    throw missing(target.logicalName)
                }
                if (target.format == CapturePlan.ArtifactFormat.PNG) validatePng(resource)
                for (output in target.outputs) {
                    bytes = Math.addExact(bytes, estimate(output, resource))
                }
            }
            return bytes
        } catch (_: ArithmeticException) {
            throw RuntimeJobExecutor.Failure(
                ErrorCode.ERROR_CODE_ARTIFACT_TOO_LARGE,
                "Artifact estimate is too large.",
            )
        }
    }

    private fun missing(name: String): RuntimeJobExecutor.Failure =
        RuntimeJobExecutor.Failure(
            ErrorCode.ERROR_CODE_RESOURCE_NOT_FOUND,
            "Capture resource was not found: $name",
        )

    private fun canonical(name: String): String = name.lowercase(Locale.ROOT)

    private fun validatePng(resource: ResourceCatalog.ResourceDescriptor) {
        val packed = setOf("RGB10_A2", "RGB10_A2UI", "R11F_G11F_B10F", "RGB9_E5", "RGB5_A1", "RGBA4")
        val reason = when {
            resource.internalFormat.contains("COMPRESSED", ignoreCase = true) -> "compressed"
            resource.numericClass.equals("stencil", ignoreCase = true) -> "stencil-only"
            (resource.numericClass.equals("sint", ignoreCase = true) ||
                resource.numericClass.equals("uint", ignoreCase = true)) && resource.componentBits > 16 ->
                "32-bit integer"
            resource.internalFormat.uppercase(Locale.ROOT) in packed -> "packed"
            resource.channelLayout !in setOf("R", "RG", "RGB", "RGBA", "DEPTH", "DEPTH_STENCIL", "") ->
                "channel layout ${resource.channelLayout}"
            else -> null
        }
        if (reason != null) throw RuntimeJobExecutor.Failure(
            ErrorCode.ERROR_CODE_CAPTURE_FAILED,
            "PNG export does not support ${resource.internalFormat} ($reason); use format=bin.",
        )
    }

    private fun estimate(
        output: CapturePlan.ArtifactOutputSpec,
        resource: ResourceCatalog.ResourceDescriptor,
    ): Long = when (output.format) {
        CapturePlan.ArtifactFormat.BIN -> resource.byteSize
        CapturePlan.ArtifactFormat.TEXT -> resource.byteSize
        CapturePlan.ArtifactFormat.JSON -> 4_096L
        CapturePlan.ArtifactFormat.PNG -> {
            val channels = when (resource.channelLayout) {
                "R", "DEPTH", "DEPTH_STENCIL" -> 1
                "RG" -> 2
                "RGB" -> 3
                "RGBA" -> 4
                else -> maxOf(1, resource.channelCount)
            }
            val bits = if (resource.numericClass.equals("float", true) ||
                resource.numericClass.startsWith("depth", true)) 16 else maxOf(8, resource.componentBits)
            val rows = Math.multiplyExact(resource.height.toLong(), 1L +
                Math.multiplyExact(resource.width.toLong(), channels.toLong() * ((bits + 7) / 8)))
            Math.addExact(rows, 1_024L)
        }
        CapturePlan.ArtifactFormat.EXR -> resource.byteSize
    }

    private fun supported(target: CapturePlan.Target): Boolean =
        when (target.kind) {
            ResourceCatalog.ResourceKind.FINAL_FRAMEBUFFER ->
                target.format == CapturePlan.ArtifactFormat.PNG
            ResourceCatalog.ResourceKind.TEXTURE ->
                target.format == CapturePlan.ArtifactFormat.PNG ||
                    target.format == CapturePlan.ArtifactFormat.BIN
            ResourceCatalog.ResourceKind.BUFFER ->
                target.format == CapturePlan.ArtifactFormat.BIN
            ResourceCatalog.ResourceKind.PATCHED_SHADERS ->
                target.format == CapturePlan.ArtifactFormat.TEXT
        }

    fun format(
        format: dev.vibris.protocol.v2.ArtifactFormat,
        fallback: CapturePlan.ArtifactFormat,
    ): CapturePlan.ArtifactFormat =
        when (format) {
            dev.vibris.protocol.v2.ArtifactFormat.ARTIFACT_FORMAT_PNG -> CapturePlan.ArtifactFormat.PNG
            dev.vibris.protocol.v2.ArtifactFormat.ARTIFACT_FORMAT_EXR -> CapturePlan.ArtifactFormat.EXR
            dev.vibris.protocol.v2.ArtifactFormat.ARTIFACT_FORMAT_BIN -> CapturePlan.ArtifactFormat.BIN
            dev.vibris.protocol.v2.ArtifactFormat.ARTIFACT_FORMAT_TEXT -> CapturePlan.ArtifactFormat.TEXT
            else -> fallback
        }

    private fun textureFormat(format: dev.vibris.protocol.v2.ArtifactFormat): CapturePlan.ArtifactFormat =
        when (format) {
            dev.vibris.protocol.v2.ArtifactFormat.ARTIFACT_FORMAT_PNG -> CapturePlan.ArtifactFormat.PNG
            dev.vibris.protocol.v2.ArtifactFormat.ARTIFACT_FORMAT_BIN -> CapturePlan.ArtifactFormat.BIN
            else -> throw RuntimeJobExecutor.Failure(
                ErrorCode.ERROR_CODE_CAPTURE_FAILED,
                "dump_texture format must be png or bin.",
            )
        }

    private fun expand(target: CapturePlan.Target, catalog: ResourceCatalog): CapturePlan.Target {
        val resource = catalog.resources
            .firstOrNull { it.kind == target.kind && it.logicalName == target.logicalName }
            ?: throw missing(target.logicalName)
        val outputs = ArrayList<CapturePlan.ArtifactOutputSpec>()
        if (target.kind == ResourceCatalog.ResourceKind.TEXTURE &&
            target.format == CapturePlan.ArtifactFormat.PNG && resource.depth > 1
        ) {
            val digits = maxOf(4, (resource.depth - 1).toString().length)
            repeat(resource.depth) { layer ->
                outputs.add(
                    CapturePlan.ArtifactOutputSpec(
                        "${target.artifactName}.layer${layer.toString().padStart(digits, '0')}.png",
                        CapturePlan.ArtifactFormat.PNG,
                        CapturePlan.ArtifactRole.SUBRESOURCE,
                        layer,
                    ),
                )
            }
        } else {
            val extension = target.format.name.lowercase(Locale.ROOT)
            val fileName = if (target.artifactName.lowercase(Locale.ROOT).endsWith(".$extension")) {
                target.artifactName
            } else {
                "${target.artifactName}.$extension"
            }
            outputs.add(
                CapturePlan.ArtifactOutputSpec(
                    fileName,
                    target.format,
                    CapturePlan.ArtifactRole.PRIMARY,
                    null,
                ),
            )
        }
        if (target.kind != ResourceCatalog.ResourceKind.FINAL_FRAMEBUFFER) {
            outputs.add(
                CapturePlan.ArtifactOutputSpec(
                    target.metadataFileName(),
                    CapturePlan.ArtifactFormat.JSON,
                    CapturePlan.ArtifactRole.METADATA,
                    null,
                ),
            )
        }
        return target.copy(outputs = outputs)
    }
}
